observer/obs_conf.go

```diff
@@ -56,12 +57,33 @@ func (c *ObsConf) validateDebugAddr() error {
 	return nil
 }
 
-func (c *ObsConf) makeMonitors() ([]*monitor, []error, error) {
+func (c *ObsConf) makeMonitors(reg *prometheus.Registerer) ([]*monitor, []error, error) {
 	var errs []error
 	var monitors []*monitor
+	collectors := make(map[string]map[string]*prometheus.Collector)
 	for e, m := range c.MonConfs {
 		entry := strconv.Itoa(e + 1)
-		monitor, err := m.makeMonitor()
+
+		kind := probers.NormalizedKind(m.Kind)
+		pCollectors, exist := collectors[kind]
+		if !exist {
+			pConf, err := probers.GetConfigurer(kind)
+			if err != nil {
+				// append validation error to errs
+				message := fmt.Errorf("'monitors' entry #%s couldn't be validated: %v", entry, err)
+				errs = append(errs, message)
+
+				// increment metrics
+				countMonitors.WithLabelValues(m.Kind, "false").Inc()
+
+				// Skip trying to call m.MakeMonitor(). We know it will fail
+				continue
+			}
+			pCollectors = pConf.Instrument()
+			collectors[kind] = pCollectors
+		}
+
+		monitor, err := m.makeMonitor(pCollectors)
```

Comment:
For this block, we want to perform all of our metric construction and registration at the top. This keeps the new logic completely separate from operations we always want to do (e.g. re-assigning `entry`, constructing the `monitor`, and incrementing `countMonitors` `false`/`true`.

```suggestion
		// setup custom metrics internal to each prober kind
		_, exists := proberSpecificMetrics[m.Kind]
		if !exists {
			// we haven't seen this prober kind before, so we need request any
			// custom metrics it may have and register them with the prometheus
			// registry
			collectors := probers.Registry[m.Kind].Instrument()
			if collectors != nil {
				for name, collector := range collectors {
					// register the collector with the metrics registry
					metrics.MustRegister(collector)
					// store the registered collector so we can pass it to every
					// monitor that will construct this kind of prober
					proberSpecificMetrics[m.Kind][name] = collector
				}
			}
		}

		entry := strconv.Itoa(e + 1)
		// Pass the mapping of prober-specific metrics to the monitor
		// constructor.
		monitor, err := m.makeMonitor(proberSpecificMetrics[m.Kind])
```